File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/LineBinderTest.java
package com.flat.simplex.parser.logic;

import com.flat.simplex.lexer.Key;
import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.logic.block.BlockIf;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.parser.logic.line.LineValue;
import com.flat.simplex.parser.logic.line.call.CallField;
import com.flat.simplex.parser.logic.line.call.CallIndexer;
import com.flat.simplex.parser.logic.line.call.CallMethod;
import com.flat.simplex.parser.logic.line.call.CallValue;
import com.flat.simplex.support.LineCallChain;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.LineCallChain.lChain;
import static com.flat.simplex.support.TokenChain.readChain;

class LineBinderTest {

    public LineValue bind(Block block, Token token, Token end) {
        LineReader reader = new LineReader(block);
        LineBinder binder = new LineBinder(block, token);
        var read = reader.read(token, end);
        return binder.bind(read);
    }

    @Test
    public void load() {
        TokenChain chain = readChain("a.b.c");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class, CallField.class, CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineOperation() {
        TokenChain chain = readChain("a + d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Add).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineDoubleOperation() {
        TokenChain chain = readChain("a + d + e");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(lChain(CallField.class).add(Key.Add).add(CallField.class)).add(Key.Add).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineDifferentOperation() {
        TokenChain chain = readChain("a+d*e");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Add).add(lChain(CallField.class).add(Key.Mul).add(CallField.class));
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineSetOperation() {
        TokenChain chain = readChain("a = e");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Set).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineDoubleSetOperation() {
        TokenChain chain = readChain("a = b = c");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Set).add(lChain(CallField.class).add(Key.Set).add(CallField.class));
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineUnaryOperation() {
        TokenChain chain = readChain("!a");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(Key.Not).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineAddUnaryOperation() {
        TokenChain chain = readChain("b + +a");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Add).add(lChain(Key.Add).add(CallField.class));
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLinePostfixOperation() {
        TokenChain chain = readChain("b++ + a");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(lChain(CallField.class).add(Key.Inc)).add(Key.Add).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineTernaryOperation() {
        TokenChain chain = readChain("a ? b : c");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Quest).add(CallField.class).add(Key.Colon).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineInnerTernaryOperation() {
        TokenChain chain = readChain("a ? a ? b : c : c");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Quest)
                .add(lChain(CallField.class).add(Key.Quest).add(CallField.class).add(Key.Colon).add(CallField.class))
                .add(Key.Colon).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineOutTernaryOperation() {
        TokenChain chain = readChain("a ? b : c ? d : e");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(lChain(CallField.class).add(Key.Quest).add(CallField.class).add(Key.Colon)
                .add(CallField.class)).add(Key.Quest).add(CallField.class).add(Key.Colon).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineInnerTernarySetOperation() {
        TokenChain chain = readChain("a ? b = d : c");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Quest)
                .add(lChain(CallField.class).add(Key.Set).add(CallField.class))
                .add(Key.Colon).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineComplexOperation() {
        TokenChain chain = readChain("e = a + b * c + d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Set)
                .add(lChain(lChain(CallField.class).add(Key.Add)
                        .add(lChain(CallField.class).add(Key.Mul).add(CallField.class))).add(Key.Add).add(CallField.class));
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadLineMissingOperator_Fail() {
        TokenChain chain = readChain("a + * d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Add).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall);
    }

    @Test
    public void loadLineTooMuchOperator_Fail() {
        TokenChain chain = readChain("a * | & % d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Mul).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall, Error.lineUnexpectedCall, Error.lineUnexpectedCall);
    }

    @Test
    public void loadLineMissingOperatorMiddle_Fail() {
        TokenChain chain = readChain("a * d *");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Mul).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall);
    }

    @Test
    public void loadLineMissingOperatorSet_Fail() {
        TokenChain chain = readChain("a = = d =");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Set).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall, Error.lineUnexpectedCall);
    }

    @Test
    public void loadLineNoOperator_Fail() {
        TokenChain chain = readChain("a !b");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall);
    }

    @Test
    public void loadLineDoubleOperator_Fail() {
        TokenChain chain = readChain("a + +");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall, Error.lineUnexpectedCall);
    }

    @Test
    public void loadIncompleteTernaryMissingLeft_Fail() {
        TokenChain chain = readChain(" ? a : b");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadIncompleteTernaryMissingQuest_Fail() {
        TokenChain chain = readChain("a : b");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadIncompleteTernaryMissingColon_Fail() {
        TokenChain chain = readChain("a ? b");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadIncompleteTernaryMissingEnd_Fail() {
        TokenChain chain = readChain("a ? b :");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadIncompleteTernaryMissingMiddle_Fail() {
        TokenChain chain = readChain("a ? : b");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadAfterTernaryOperator_Fail() {
        TokenChain chain = readChain("a ? b : : d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Quest).add(CallField.class).add(Key.Colon).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall);
    }

    @Test
    public void loadTernaryEmptyResultInside_Fail() {
        TokenChain chain = readChain("a ? * : d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadTernaryDoubleResultInside_Fail() {
        TokenChain chain = readChain("a ? b c : d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineMissingAccessor, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadTernarySetterInside_Fail() {
        TokenChain chain = readChain("a ? = : d");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class).add(Key.Set).add(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadSetAfterIncrement_Fail() {
        TokenChain chain = readChain("++indexer = 10");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(lChain(Key.Inc).add(CallField.class)).add(Key.Set).add(CallValue.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineSetOperator);
    }

    @Test
    public void loadSetIndexer() {
        TokenChain chain = readChain("indexer[5] = 10");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class, CallIndexer.class).add(Key.Set).add(CallValue.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context);
    }

    @Test
    public void loadSetAfterIndexerIncrement_Fail() {
        TokenChain chain = readChain("++indexer[5] = 10");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(lChain(Key.Inc).add(CallField.class, CallIndexer.class)).add(Key.Set).add(CallValue.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineSetOperator);
    }

    @Test
    public void loadSetMethod_Fail() {
        TokenChain chain = readChain("method() = 10");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class, CallMethod.class).add(Key.Set).add(CallValue.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineSetOperator);
    }

    @Test
    public void loadSetValue_Fail() {
        TokenChain chain = readChain("20 = 10");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallValue.class).add(Key.Set).add(CallValue.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineSetOperator);
    }

    @Test
    public void loadMultiplyNothing_Fail() {
        TokenChain chain = readChain("a *");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = lChain(CallField.class);
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall);
    }

    @Test
    public void loadMultiplyAlone_Fail() {
        TokenChain chain = readChain("*");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = new LineCallChain();
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.lineUnexpectedCall);
    }

    @Test
    public void loadEmptyFromReader_Fail() {
        TokenChain chain = readChain("$");

        Context context = new Context(chain.get());
        BlockIf block = getBlock(context);
        LineValue line = bind(block, chain.get(), null);

        LineCallChain callChain = new LineCallChain();
        callChain.assertChain(line, "Invalid line chain");
        assertErrors(context, Error.unexpectedToken, Error.lineEmptyLine);
    }

    private BlockIf getBlock(Context context) {
        TokenChain chain = readChain("if(true);");

        return new BlockIf(context, chain.get(), null);

    }
}

File: SimplexApp/simplex/parser/logic/LineBinder.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const LineCallChain = require("simplex/support/LineCallChain.js");
const LineReader = require("simplex/parser/logic/LineReader.js");
const LineBinder = require("simplex/parser/logic/LineBinder.js");
const Context = require("simplex/parser/logic/Context.js");
const BlockIf = require("simplex/parser/logic/block/BlockIf.js");
const ContextSupport = require("simplex/support/ContextSupport.js");
const Key = require("simplex/lexer/Key.js");
const Error = require("simplex/parser/logic/error/Error.js");

const readChain = TokenChain.readChain;
const lChain = LineCallChain.lChain;
const assertErrors = ContextSupport.assertErrors;

function bind(block, token, end) {
    let reader = new LineReader(block);
    let binder = new LineBinder(block, token);
    var read = reader.read(token, end);
    return binder.bind(read);
}

function getBlock(context) {
    let chain = readChain("if(true);");

    return new BlockIf(context, chain.get(), null);
}

test('load', () => {
    let chain = readChain("a.b.c");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField', 'CallField', 'CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineOperation', () => {
    let chain = readChain("a + d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Add).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineDoubleOperation', () => {
    let chain = readChain("a + d + e");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain(lChain('CallField').add(Key.Add).add('CallField')).add(Key.Add).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineDifferentOperation', () => {
    let chain = readChain("a+d*e");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Add).add(lChain('CallField').add(Key.Mul).add('CallField'));
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineSetOperation', () => {
    let chain = readChain("a = e");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Set).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineDoubleSetOperation', () => {
    let chain = readChain("a = b = c");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Set).add(lChain('CallField').add(Key.Set).add('CallField'));
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineUnaryOperation', () => {
    let chain = readChain("!a");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain(Key.Not).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineAddUnaryOperation', () => {
    let chain = readChain("b + +a");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Add).add(lChain(Key.Add).add('CallField'));
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLinePostfixOperation', () => {
    let chain = readChain("b++ + a");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain(lChain('CallField').add(Key.Inc)).add(Key.Add).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineTernaryOperation', () => {
    let chain = readChain("a ? b : c");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Quest).add('CallField').add(Key.Colon).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineInnerTernaryOperation', () => {
    let chain = readChain("a ? a ? b : c : c");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Quest)
        .add(lChain('CallField').add(Key.Quest).add('CallField').add(Key.Colon).add('CallField'))
        .add(Key.Colon).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineOutTernaryOperation', () => {
    let chain = readChain("a ? b : c ? d : e");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain(lChain('CallField').add(Key.Quest).add('CallField').add(Key.Colon)
        .add('CallField')).add(Key.Quest).add('CallField').add(Key.Colon).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineInnerTernarySetOperation', () => {
    let chain = readChain("a ? b = d : c");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Quest)
        .add(lChain('CallField').add(Key.Set).add('CallField'))
        .add(Key.Colon).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineComplexOperation', () => {
    let chain = readChain("e = a + b * c + d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Set)
        .add(lChain(lChain('CallField').add(Key.Add)
            .add(lChain('CallField').add(Key.Mul).add('CallField'))).add(Key.Add).add('CallField'));
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadLineMissingOperator_Fail', () => {
    let chain = readChain("a + * d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Add).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall);
});

test('loadLineTooMuchOperator_Fail', () => {
    let chain = readChain("a * | & % d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Mul).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall, Error.lineUnexpectedCall, Error.lineUnexpectedCall);
});

test('loadLineMissingOperatorMiddle_Fail', () => {
    let chain = readChain("a * d *");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Mul).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall);
});

test('loadLineMissingOperatorSet_Fail', () => {
    let chain = readChain("a = = d =");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Set).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall, Error.lineUnexpectedCall);
});

test('loadLineNoOperator_Fail', () => {
    let chain = readChain("a !b");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall);
});

test('loadLineDoubleOperator_Fail', () => {
    let chain = readChain("a + +");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall, Error.lineUnexpectedCall);
});

test('loadIncompleteTernaryMissingLeft_Fail', () => {
    let chain = readChain(" ? a : b");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineTernaryIncomplete);
});

test('loadIncompleteTernaryMissingQuest_Fail', () => {
    let chain = readChain("a : b");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineTernaryIncomplete);
});

test('loadIncompleteTernaryMissingColon_Fail', () => {
    let chain = readChain("a ? b");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineTernaryIncomplete);
});

test('loadIncompleteTernaryMissingEnd_Fail', () => {
    let chain = readChain("a ? b :");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineTernaryIncomplete);
});

test('loadIncompleteTernaryMissingMiddle_Fail', () => {
    let chain = readChain("a ? : b");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall, Error.lineTernaryIncomplete);
});

test('loadAfterTernaryOperator_Fail', () => {
    let chain = readChain("a ? b : : d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Quest).add('CallField').add(Key.Colon).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall);
});

test('loadTernaryEmptyResultInside_Fail', () => {
    let chain = readChain("a ? * : d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall, Error.lineTernaryIncomplete);
});

test('loadTernaryDoubleResultInside_Fail', () => {
    let chain = readChain("a ? b c : d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineMissingAccessor, Error.lineTernaryIncomplete);
});

test('loadTernarySetterInside_Fail', () => {
    let chain = readChain("a ? = : d");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField').add(Key.Set).add('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineTernaryIncomplete);
});

test('loadSetAfterIncrement_Fail', () => {
    let chain = readChain("++indexer = 10");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain(lChain(Key.Inc).add('CallField')).add(Key.Set).add('CallValue');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineSetOperator);
});

test('loadSetIndexer', () => {
    let chain = readChain("indexer[5] = 10");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField', 'CallIndexer').add(Key.Set).add('CallValue');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context);
});

test('loadSetAfterIndexerIncrement_Fail', () => {
    let chain = readChain("++indexer[5] = 10");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain(lChain(Key.Inc).add('CallField', 'CallIndexer')).add(Key.Set).add('CallValue');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineSetOperator);
});

test('loadSetMethod_Fail', () => {
    let chain = readChain("method() = 10");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField', 'CallMethod').add(Key.Set).add('CallValue');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineSetOperator);
});

test('loadSetValue_Fail', () => {
    let chain = readChain("20 = 10");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallValue').add(Key.Set).add('CallValue');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineSetOperator);
});

test('loadMultiplyNothing_Fail', () => {
    let chain = readChain("a *");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = lChain('CallField');
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall);
});

test('loadMultiplyAlone_Fail', () => {
    let chain = readChain("*");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = new LineCallChain();
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.lineUnexpectedCall);
});

test('loadEmptyFromReader_Fail', () => {
    let chain = readChain("$");

    let context = new Context(chain.get());
    let block = getBlock(context);
    let line = bind(block, chain.get(), null);

    let callChain = new LineCallChain();
    callChain.assertChain(line, "Invalid line chain");
    assertErrors(context, Error.unexpectedToken, Error.lineEmptyLine);
});